File: 2121-find-if-path-exists-in-graph/find-if-path-exists-in-graph.java
class Solution {
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        // Step 1: Build adjacency list
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }

        // Add each edge to the graph (undirected)
        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];
            graph.get(u).add(v);
            graph.get(v).add(u);
        }

        // Step 2: Create visited array
        boolean[] visited = new boolean[n];

        // Step 3: DFS to check if a path exists
        return dfs(graph, source, destination, visited);
    }

    private boolean dfs(List<List<Integer>> graph, int current, int destination, boolean[] visited) {
        if (current == destination) return true; // base case

        visited[current] = true; // mark current as visited

        // Explore neighbors
        for (int neighbor : graph.get(current)) {
            if (!visited[neighbor]) {
                if (dfs(graph, neighbor, destination, visited)) {
                    return true;
                }
            }
        }

        return false; // No path found
    }
}

File: 2121-find-if-path-exists-in-graph/find-if-path-exists-in-graph.js
var validPath = function(n, edges, source, destination) {
    const adj = Array.from({length: n}, () => []);
    for (let [u, v] of edges) {
        adj[u].push(v);
        adj[v].push(u); // undirected
    }

    const visited = new Array(n).fill(false);
    const queue = [source];
    visited[source] = true;

    while (queue.length) {
        const node = queue.shift();
        if (node === destination) return true;

        for (let neighbor of adj[node]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                queue.push(neighbor);
            }
        }
    }

    return false;
};
